tests/test_flow/fakes.py

```diff
@@ -56,3 +65,164 @@ def submit(self, task_config, fn, *args, **kwargs):
             task_config=task_config,
             function=partial(fn, *args, **kwargs),
         ).submit()
+
+
+class FakeBlob:
+    def __init__(self, bucket, name, blobs):
+        assert not name.startswith("/")
+        self.bucket = bucket
+        self.name = name
+        self.blobs = blobs
+
+    def upload_from_string(self, data):
+        self.blobs[self._url] = data
+
+    def download_as_string(self):
+        return self.blobs[self._url]
+
+    def upload_from_filename(self, filename):
+        self.upload_from_string(open(filename, "r").read())
+
+    def download_to_filename(self, filename):
+        open(filename, "w").write(self.download_as_string())
+
+    def exists(self):
+        return self._url in self.blobs
+
+    def delete(self):
+        del self.blobs[self._url]
+
+    @property
+    def _url(self):
+        return f"gs://{self.bucket}/{self.name}"
+
+
+class FakeBucket:
+    def __init__(self, name, blobs):
+        self.name = name
+        self.blobs = blobs
+
+    def blob(self, blob_name):
+        return FakeBlob(self.name, blob_name, self.blobs)
+
+    def get_blob(self, blob_name):
+        return self.blob(blob_name)
+
+    def list_blobs(self, prefix):
+        assert not prefix.startswith("gs://")
+        return [
+            FakeBlob(self.name, k.replace(f"gs://{self.name}/", "", 1), self.blobs)
+            for (k, v) in self.blobs.items()
+            if k.startswith(f"gs://{self.name}/{prefix}")
+        ]
+
+
+class FakeGCS:
+    def __init__(self):
+        self.blobs = {}
+
+    def get_bucket(self, bucket):
+        return FakeBucket(bucket, self.blobs)
+
+    def wipe_path(self, url):
+        urls_to_remove = [k for k in self.blobs.keys() if k.startswith(url)]
+        for k in urls_to_remove:
+            del self.blobs[k]
+
+    # Allows the blob to be used like a file object.
+    # Used for mocking out blocks.filesystem.GCSNativeFileSystem.
+    @contextmanager
+    def open_blob(self, path, mode):
+        f = io.BytesIO(self.blobs.get(path, b""))
+        yield f
+        self.blobs[path] = f.getvalue()
+
+    def block_pickle(self, result, path):
+        self.blobs[path] = result
+
+    def block_unpickle(self, path):
+        return self.blobs[path]
+
+    def gsutil_cp(self, src_url, dst_url):
+        """
+        Emulate gsutil with the cp command.
+        """
```

Comment:
Turns out emulating `gsutil cp` is not straightforward. There are slightly different logic depending on whether the source / destination url is a file or directory.